openpype/hooks/pre_copy_last_published_workfile.py

```diff
@@ -0,0 +1,169 @@
+            "use_last_published_workfile"
+        )
+        if use_last_published_workfile is None:
+            self.log.info(
+                (
+                    "Seems like old version of settings is used."
+                    ' Can\'t access custom templates in host "{}".'
+                ).format(host_name)
+            )
+            return
+        elif use_last_published_workfile is False:
+            self.log.info(
+                (
+                    'Project "{}" has turned off to use last published'
+                    ' workfile as first workfile for host "{}"'
+                ).format(project_name, host_name)
+            )
+            return
+
+        self.log.info("Trying to fetch last published workfile...")
+
+        # Check there is no workfile available
+        last_workfile = self.data.get("last_workfile_path")
+        if os.path.exists(last_workfile):
+            self.log.debug(
+                "Last workfile exists. Skipping {} process.".format(
+                    self.__class__.__name__
+                )
+            )
+            return
+
+        project_doc = self.data.get("project_doc")
+        asset_doc = self.data.get("asset_doc")
+        anatomy = self.data.get("anatomy")
+        if project_doc and asset_doc:
+            # Get subset id
+            subset_id = next(
+                (
+                    subset["_id"]
+                    for subset in get_subsets(
+                        project_name,
+                        asset_ids=[asset_doc["_id"]],
+                        fields=["_id", "data.family"],
+                    )
+                    if subset["data"]["family"] == "workfile"
+                ),
+                None,
+            )
+            if not subset_id:
+                self.log.debug('No any workfile for asset "{}".').format(
+                    asset_doc["name"]
+                )
+                return
+
+            # Get workfile representation
+            workfile_representation = next(
+                (
+                    representation
+                    for representation in get_representations(
+                        project_name,
+                        version_ids=[
+                            get_last_version_by_subset_id(
+                                project_name, subset_id, fields=["_id"]
+                            )["_id"]
+                        ],
+                    )
+                    if representation["context"]["task"]["name"] == task_name
+                ),
+                None,
+            )
+
+            if not workfile_representation:
+                self.log.debug(
+                    'No published workfile for task "{}" and host "{}".'
+                ).format(task_name, host_name)
+                return
+
+            # Get sync server from Tray,
+            # which handles the asynchronous thread instance
+            sync_server = next(
```

Comment:
I would recommend not using this global variables approach. There may be multiple created instances of `ModulesManager` at the time this is running. Also consider that prelaunch hooks would be launched in separated process from the main thread (which is in plan). Please consider using rest api endpoint for local webserver rather then using this approach.

Reply:
I understand but I cannot figure how to wait for the files copy to finish if I cannot access the asynchronous thread. If I start a `while`, it'll stop the futures stack from proceeding. 

I may try to get memory pointers from the REST API but I'm not sure how it could work.

Does @kalisp have any inputs regarding this issue?

Reply:
It would require that sync server has 2 rest points:
- to "trigger sync" of representation
- to "check" if is representation synced

Reply:
Yes that's what I'd have done, but it doesn't solve the concurrency issue. The server is built on an asynchronous model and not on a multiprocessing system. It means if something is running on the main thread, it'll block the server to execute anything. And that's waiting for a "check" will do. If I create a whole new asynchronous loop in the hook, the rest of the hooks will proceed and the file will be open before it's been downloaded.

Reply:
What about checking DB (periodically). It should fill 'created_dt' for particular site when it is finished.

Reply:
In fact, I was wrong, the implementation I described works fine. Hope it suits your requirements better.